### ISSUE TITLE
Sensor controller panics if sensor is created after the watched resource

### ISSUE THREAD

**Issue by shrinandj** [argoproj/argo-events#84]
**Describe the bug**
The sensor-controller panicked when a resource that the sensor depended upon already existed. More below...

**To Reproduce**
1. I created a sensor that was supposed to run an Argo workflow in response to a Kubernetes secret.
```
$ cat examples/resource-sensor-secret.yaml
apiVersion: argoproj.io/v1alpha1
kind: Sensor
metadata:
  generateName: resource-example-
  labels:
    sensors.argoproj.io/controller-instanceid: axis
spec:
  signals:
    - name: secret-create
      resource:
        namespace: test
        group: ""
        version: "v1"
        kind: "Secret"
        filter:
          prefix: my-secret
  triggers:
    - name: ns-workflow
      resource:
        namespace: default
        group: argoproj.io
        version: v1alpha1
        kind: Workflow
        source:
          inline: |
              apiVersion: argoproj.io/v1alpha1
              kind: Workflow
              metadata:
                generateName: hello-world-
              spec:
                entrypoint: whalesay
                templates:
                  -
                    container:
                      args:
                        - "hello world"
                      command:
                        - cowsay
                      image: "docker/whalesay:latest"
                    name: whalesay
```

2. When the above sensor was created, the namespace `test` (and therefore the secret `my-secret` in it) did not exist.

3. Then I created the `test` namespace and `my-secret` in it.

4. The sensor correctly fired the trigger and the Argo workflow executed.

5. After this, I create a second sensor identical to the above (`kubectl create -f examples/resource-sensor-secret.yaml`).

6. This cause the sensor-controller to panic.

```
time="2018-08-08T22:06:54Z" level=error msg="runtime error: invalid memory address or nil pointer dereference" namespace=default sensor=resource-example-zz7lf
time="2018-08-08T22:06:54Z" level=error msg="recovered from panic: runtime error: invalid memory address or nil pointer dereference\ngoroutine 45 [running]:\nruntime/debug.Stack(0xc420305220, 0xc4206698b8, 
0x1)\n\t/usr/local/Cellar/go/1.10.3/libexec/src/runtime/debug/stack.go:24 +0xa7\ngithub.com/argoproj/argo-events/controller.(*sOperationCtx).operate.func1(0xc4200129a0)\n\t/Users/sjavadekar/ws/go/src/github.com/argoproj/argo-
events/controller/operator.go:70 +0x95\npanic(0x13a5de0, 0x1fa7950)\n\t/usr/local/Cellar/go/1.10.3/libexec/src/runtime/panic.go:502 +0x229\ngithub.com/argoproj/argo-events/controller.(*sOperationCtx).processSignal(0xc4200129a0, 0xc4204ce0c0, 0xd, 
0x0, 0x0, 0x0, 0x0, 0xc420305180, 0x0, 0x0, ...)\n\t/Users/sjavadekar/ws/go/src/github.com/argoproj/argo-events/controller/signal.go:77 +0x2f0\ngithub.com/argoproj/argo-events/controller.(*sOperationCtx).operate(0xc4200129a0, 0x0, 
0x0)\n\t/Users/sjavadekar/ws/go/src/github.com/argoproj/argo-events/controller/operator.go:88 +0x1be\ngithub.com/argoproj/argo-events/controller.(*SensorController).processNextItem(0xc42014e640, 
0xc420134600)\n\t/Users/sjavadekar/ws/go/src/github.com/argoproj/argo-events/controller/controller.go:118 +0x21d\ngithub.com/argoproj/argo-events/controller.
(*SensorController).runWorker(0xc42014e640)\n\t/Users/sjavadekar/ws/go/src/github.com/argoproj/argo-events/controller/controller.go:182 +0x2b\ngithub.com/argoproj/argo-events/controller.(*SensorController).(github.com/argoproj/argo-
events/controller.runWorker)-fm()\n\t/Users/sjavadekar/ws/go/src/github.com/argoproj/argo-events/controller/controller.go:175 +0x2a\ngithub.com/argoproj/argo-
events/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc4204f9b20)\n\t/Users/sjavadekar/ws/go/src/github.com/argoproj/argo-events/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x54\ngithub.com/argoproj/argo-
events/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc4204f9b20, 0x3b9aca00, 0x0, 0x1, 0x0)\n\t/Users/sjavadekar/ws/go/src/github.com/argoproj/argo-events/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xbd\ngithub.com/argoproj/argo-
events/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc4204f9b20, 0x3b9aca00, 0x0)\n\t/Users/sjavadekar/ws/go/src/github.com/argoproj/argo-events/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d\ncreated by github.com/argoproj/argo-
events/controller.(*SensorController).Run\n\t/Users/sjavadekar/ws/go/src/github.com/argoproj/argo-events/controller/controller.go:175 +0x2f8\n" namespace=default sensor=resource-example-zz7lf

```

**Expected behavior**
Not sure what is expected here.

1. Should the second sensor fire the trigger beacuse the secret `my-secret` exists?
2. Maybe the second sensor should simply wait forever since the secret `my-secret` already exists (and the actual event of creation of the secret is already past)

**Comment by shrinandj** [comment]
Seems like the problematic code is: `github.com/argoproj/argo-events/controller/signal.go:77` which is:
```
func (soc *sOperationCtx) processSignal(signal v1alpha1.Signal) (*v1alpha1.NodeStatus, error) {
	soc.log.Debugf("evaluating signal '%s'", signal.Name)
...
	if node.LatestEvent != nil {
		if !node.LatestEvent.Seen {
			soc.s.Status.Nodes[node.ID].LatestEvent.Seen = true .  <<<------------------------
			soc.updated = true
		}
...
```

**Comment by shrinandj** [comment]
The actual object does not seem to have an ID. As a result `node.ID` is an empty string causing the panic.

```
...
status:
  completedAt: 2018-08-08T22:42:17Z
  message: 'runtime error: invalid memory address or nil pointer dereference'
  nodes:
    resource-example-dbzdm-1158636224:
      completedAt: null
      displayName: ""
      id: ""
```

**Comment by magaldima** [comment]
can you include more logs before the error? it looks like the sensor may not be persisted successfully after an initial `operate`. it's also odd that this only happens after an identical sensor was created... i can't imagine the `nodeID` could be the same for both sensors since you're using `generateName` and those hashes would be different.

**Comment by shrinandj** [comment]
```
time="2018-08-09T19:57:14Z" level=info msg="sensor controller (version: v0.1.2+38548a7.dirty) (instance: axis) starting"
time="2018-08-09T19:57:14Z" level=info msg="watching sensor controller config map updates"
time="2018-08-09T19:57:14Z" level=info msg="detected ConfigMap update. updating the controller config."
time="2018-08-09T20:00:31Z" level=info msg="Signal 'secret-create' initialized: " namespace=default sensor=resource-example-vmd6s
time="2018-08-09T20:00:31Z" level=info msg="Signal 'secret-create' phase  -> Active" namespace=default sensor=resource-example-vmd6s
time="2018-08-09T20:00:31Z" level=info msg="Signal 'secret-create' message  -> stream established" namespace=default sensor=resource-example-vmd6s
time="2018-08-09T20:00:31Z" level=info msg="sensor phase  -> Active" namespace=default sensor=resource-example-vmd6s
time="2018-08-09T20:00:31Z" level=info msg="sensor message  -> listening for signal events" namespace=default sensor=resource-example-vmd6s
time="2018-08-09T20:00:31Z" level=error msg="Event Stream (resource-example-vmd6s/secret-create) Fatal: 'resource-example-vmd6s-1158636224' node is missing from sensor's nodes"
time="2018-08-09T20:00:31Z" level=info msg="Event Stream (resource-example-vmd6s/secret-create) Msg: (Action:ACCEPTED) - Context: {   nil  2018-08-09 20:00:31 +0000 UTC nil  map[]}"
time="2018-08-09T20:00:32Z" level=info msg="sensor phase  -> Error" namespace=default sensor=resource-example-vmd6s
time="2018-08-09T20:00:32Z" level=info msg="sensor message  -> runtime error: invalid memory address or nil pointer dereference" namespace=default sensor=resource-example-vmd6s
time="2018-08-09T20:00:32Z" level=info msg="marking sensor complete" namespace=default sensor=resource-example-vmd6s
time="2018-08-09T20:00:32Z" level=error msg="runtime error: invalid memory address or nil pointer dereference" namespace=default sensor=resource-example-vmd6s
time="2018-08-09T20:00:32Z" level=error msg="recovered from panic: runtime error: invalid memory address or nil pointer dereference\ngoroutine 39 [running]:\nruntime/debug.Stack(0xc4204acc30, 0xc42071b848, 0x1)\n\t/usr/local/Cellar/go/1.10.3/libexec/src/runtime/debug/stack.go:24 +0xa7\ngithub.com/argoproj/argo-events/controller.(*sOperationCtx).operate.func1(0xc4201cb1a0)\n\t/Users/sjavadekar/ws/go/src/github.com/argoproj/argo-events/controller/operator.go:70 +0x95\npanic(0x13a5de0, 0x1fa7950)\n\t/usr/local/Cellar/go/1.10.3/libexec/src/runtime/panic.go:502 +0x229\ngithub.com/argoproj/argo-events/controller.(*sOperationCtx).processSignal(0xc4201cb1a0, 0xc4206a3af3, 0xd, 0x0, 0x0, 0x0, 0x0, 0xc4204acbe0, 0x0, 0x0, ...)\n\t/Users/sjavadekar/ws/go/src/github.com/argoproj/argo-events/controller/signal.go:99 +0x433\ngithub.com/argoproj/argo-events/controller.(*sOperationCtx).operate(0xc4201cb1a0, 0x0, 0x0)\n\t/Users/sjavadekar/ws/go/src/github.com/argoproj/argo-events/controller/operator.go:88 +0x1be\ngithub.com/argoproj/argo-events/controller.(*SensorController).processNextItem(0xc4200adea0, 0xc42009e300)\n\t/Users/sjavadekar/ws/go/src/github.com/argoproj/argo-events/controller/controller.go:118 +0x21d\ngithub.com/argoproj/argo-events/controller.(*SensorController).runWorker(0xc4200adea0)\n\t/Users/sjavadekar/ws/go/src/github.com/argoproj/argo-events/controller/controller.go:182 +0x2b\ngithub.com/argoproj/argo-events/controller.(*SensorController).(github.com/argoproj/argo-events/controller.runWorker)-fm()\n\t/Users/sjavadekar/ws/go/src/github.com/argoproj/argo-events/controller/controller.go:175 +0x2a\ngithub.com/argoproj/argo-events/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc420675070)\n\t/Users/sjavadekar/ws/go/src/github.com/argoproj/argo-events/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x54\ngithub.com/argoproj/argo-events/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc420675070, 0x3b9aca00, 0x0, 0x1, 0x0)\n\t/Users/sjavadekar/ws/go/src/github.com/argoproj/argo-events/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xbd\ngithub.com/argoproj/argo-events/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc420675070, 0x3b9aca00, 0x0)\n\t/Users/sjavadekar/ws/go/src/github.com/argoproj/argo-events/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d\ncreated by github.com/argoproj/argo-events/controller.(*SensorController).Run\n\t/Users/sjavadekar/ws/go/src/github.com/argoproj/argo-events/controller/controller.go:175 +0x2f8\n" namespace=default sensor=resource-example-vmd6s
```

**Comment by magaldima** [comment]
this looks like the signal is receiving an event on the stream before the controller has persisted the updated sensor state. we also are missing a `continue` after this first error log in the `listenOnStream` function. this should fix the nil pointer dereference. As for the fact that this is triggering immediately, we should investigate the internals of watch because if this secret was created in the past, we shouldn't be getting this event. I predict the watch is firing because it simply found a matching resource. Maybe we need to restrict based on the watch `EventType` of:
```
Added    EventType = "ADDED"
Modified EventType = "MODIFIED"
Deleted  EventType = "DELETED"
Error    EventType = "ERROR"
```
We should add this eventType to the eventContext that we return and look at the values we get for the first and second sensor here.

**Comment by VaibhavPage** [comment]
Addressed in PR #100 .
